packages/consent/consent-tools/src/domain/create-wrapper.ts

```diff
@@ -0,0 +1,186 @@
+import {
+  Categories,
+  CreateWrapper,
+  AnyAnalytics,
+  CDNSettingsIntegrations,
+  CDNSettingsRemotePlugin,
+  InitOptions,
+  CreateWrapperSettings,
+  CDNSettings,
+} from '../types'
+import { validateCategories, validateOptions } from './validation'
+import { createConsentStampingMiddleware } from './consent-stamping'
+import { pipe, pick, uniq } from '../utils'
+import { AbortLoadError, LoadContext } from './load-cancellation'
+
+export const createWrapper: CreateWrapper = (createWrapperOptions) => {
+  const {
+    shouldDisableSegment,
+    shouldDisableConsentRequirement,
+    getCategories,
+    shouldLoad,
+    integrationCategoryMappings,
+    shouldEnableIntegration,
+  } = createWrapperOptions
+
+  validateOptions(createWrapperOptions)
+
+  return (analytics) => {
+    const ogLoad = analytics.load
+
+    const loadWithConsent: AnyAnalytics['load'] = async (
+      settings,
+      options
+    ): Promise<void> => {
+      // do not load anything -- segment included
+      if (await shouldDisableSegment?.()) {
+        return
+      }
+
+      const consentRequirementDisabled =
+        await shouldDisableConsentRequirement?.()
+      if (consentRequirementDisabled) {
+        // ignore consent -- just call analytics.load as usual
+        return ogLoad.call(analytics, settings, options)
+      }
+
+      // use these categories to disable/enable the appropriate device mode plugins
+      let initialCategories: Categories
+      try {
+        initialCategories =
+          (await shouldLoad?.(new LoadContext())) || (await getCategories())
+      } catch (e: unknown) {
+        // consumer can call ctx.cancel({ loadSegmentNormally: true })
+        // to load Segment but disable consent requirement
+        if (e instanceof AbortLoadError) {
+          if (e.loadSegmentNormally === true) {
+            ogLoad.call(analytics, settings, options)
+          }
+        } else {
+          console.error(e)
+        }
+
+        return
+      }
+
+      validateCategories(initialCategories)
+
+      const cdnSettingsP = new Promise<CDNSettings>((resolve) =>
+        analytics.on('initialize', resolve)
+      )
+
+      // we don't want to send _every_ category to segment, only the ones that the user has explicitly configured in their integrations
+      const getFilteredSelectedCategories = async (): Promise<
+        Categories | undefined
+      > => {
+        const cdnSettings = await cdnSettingsP
+        let allCategories: string[]
```

Comment:
Do `allCategories` get recalculated on every event through the middleware? Since this doesn't currently change (outside of a page reload), can we cache the result after the 1st time we calculate it?

Reply:
It is only used once, but IMO it's not worth adding the complexity of state -- this is a simple O(N) transformation, and it's only relevant in the minority of cases where someone is using hardcoded categories.